### DISCUSSION TITLE
Allow middleware matcher to match route group

### DISCUSSION THREAD

**Discussion by ibrahim-ahmed-dev** [vercel/next.js#44635] [Ideas]
### Describe the feature you'd like to request

It would be nice if we could configure the middleware matcher to match an entire route group so that we can have greater flexibility with middleware.

The use cases for this could range from authorization, authentication, feature flags, etc.

### Describe the solution you'd like

Given the following example file structure
```
app
    /(unauthenticated)
        /login
            page.tsx
        /signup
            page.tsx
    /(authenticated)
        /dashboard
            page.tsx
        /account
            page.tsx
        /cart
            page.tsx
```

Allow some kind of pattern matching in the middleware config
```
export const config = {
	matcher: ["/(authenticated)/:path*"],
}
```

In this example, I could redirect the user to the login page if they are not authenticated via the middleware.

### Describe alternatives you've considered

I have considered adding a nested layout under the (authenticated) route group, however, adding a layout for the purpose of performing logic on the request seems like something that is better fitted in the middleware. Since middleware executes before the request is processed, it might result in a smoother transition to have the logic performed there.

**Comment by vladmiller** [comment]
In our use-case, something like that would allow extracting authentication logic into separate packages for easier re-use. Also, would have been wonderful, to be able to render custom components as pages.

E.g. imagine, you're adding `withAuthentication(...)` middleware. Would be great if it automatically exposes API methods like `/api/login`, `/api/profile` as well as pages like `/login`, `/password-reset`,  etc.

**Comment by bdthombre** [comment]
Yes, this would make it so easy to manage middleware matchers..

**Reply by bdthombre** [reply]
Is this request getting ignored..

**Comment by dima-kov** [comment]
+1

**Comment by Sreesanth46** [comment]
Is this feature available?

**Comment by sharfudeen** [comment]
+1

**Comment by marinofranz** [comment]
+1

**Comment by hyjoong** [comment]
It seems that, at the moment, specifying paths to be excluded and returning null is not the ideal solution, but it appears to be the best option available.

```javascript
// middleware.ts

const unauthenticatedRoutes = ['/login','signup']
const isUnauthenticated = unauthenticatedRoutes.includes(nextUrl.pathname);
if (isUnauthenticated) return null;

```

**Comment by zdavis** [comment]
+1

**Comment by wildanrhmn** [comment]
+1

**Comment by anindosarker** [comment]
I was gonna open this exact issue. For the readers, this is how I currently do the implementation

- I have a role wise page access list. 
- In the middleware, I check for the allowed paths for a role. 
- I need to add the paths in the matcher. 

example of `middleware.ts`:
```ts

import { NextRequestWithAuth, withAuth } from "next-auth/middleware";
import { NextResponse } from "next/server";
const roleAccessConfig = {
  admin: ["/"],
  tutor: ["/tutor", "/onboarding"],
  student: ["/student", "/onboarding"],
  family: ["/family", "/onboarding"],
  user: ["/onboarding"],
};

function authMiddleware(req: NextRequestWithAuth) {
  // Role based access control
  const userRole = req.nextauth.token?.user?.role;
  const pathName = req.nextUrl.pathname;

  if (userRole) {
    // Check if the user's role has access to the current page
    const allowedRoutes = roleAccessConfig[userRole as keyof typeof roleAccessConfig];
    if (!allowedRoutes || !allowedRoutes.some((route) => pathName.includes(route))) {
      return NextResponse.redirect(new URL("/unauthorized", req.nextUrl));
    }
  }
}

export default withAuth(authMiddleware, {
  callbacks: {
    authorized: async ({ token }) => {
      return !!token;
    },
  },
});

// config to exclude auth pages from being protected
export const config = {
  matcher: ["/dashboard/:path*",    "/admin/:path*"],
};
```

**Comment by santivdt** [comment]
+1

**Comment by ackoppel** [comment]
+1

**Comment by Pyakz** [comment]
+1

**Comment by jbsiddall** [comment]
+1

**Comment by docmg** [comment]
+1

I am currently using a similar workaround/implementation as anindosarker. The possibility to access the group in the middleware would have made the configuration much clearer.

**Comment by aryasaatvik** [comment]
+1

**Comment by nezilkeyvalue** [comment]
+1

**Comment by joshp-fat-potato-uk** [comment]
+1

**Comment by adeemmm23** [comment]
+1 please

**Comment by emarazz** [comment]
+1 please

**Comment by wilsondefreitas-dev** [comment]
+1 please 🙏

**Comment by cotocoopman** [comment]
+1 pls

**Comment by Wonkers0** [comment]
+1

**Comment by GBvanDam** [comment]
Would it maybe even be stronger to enable multiple middlewares in a file-system based nested manner? This way, not only route groups are added, but several more things?

https://github.com/vercel/next.js/discussions/69866#discussion-7160165

**Comment by jvzaniolo** [comment]
+1

**Comment by GBvanDam** [comment]
Lee Robinson pointed towards experimental interceptors https://github.com/vercel/next.js/pull/70961

these might answer at least part of this discussion?

**Comment by Dezmonter** [comment]
+

**Comment by r0bdiabl0** [comment]
+1

**Comment by r0bdiabl0** [comment]
This seems like a major miss. How can we not specify route groups in the middleware to cover all paths and pages in a route group. Currently, the only way to make this work is to specify each path individually which is not scalable. At first, I thought I could create a middleware group for each route group and just realized that it wasn't working because you can only have one middleware in the root.

There are going to be a lot of developers who think it's best to use root paths versus sub-directory based paths like `/members` and `/admin` as long as they can keep the page names unique. Why offer route groups then if we can't take advantage of their true purpose?

**Comment by TVScoundrel** [comment]
+1

**Comment by TVScoundrel** [comment]
You could drop middleware and use a Layout server component in the root of your route group to perform functions that the middleware would otherwise perform. That way you have granular control over all groups.

```javascript
// /app/(protected)/layout.tsx
import { getAuthCurrentUser } from "@/utils/utils";
import { headers } from "next/headers";
import { redirect } from "next/navigation";

export default async function ProtectedLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const user = await getAuthCurrentUser();

  // Get the current path from headers
  const headersList = headers();
  const currentPath = headersList.get("x-current-path") || "/";

  if (!user) {
    redirect(`/signup?source=${currentPath}`);
  }

  return <>{children}</>;
}
```

To make this work, I do have a middleware to add "x-current-path" to my headers:

```javascript
// middleware.ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

export function middleware(request: NextRequest) {
  const headers = new Headers(request.headers);
  headers.set("x-current-path", request.nextUrl.pathname);
  return NextResponse.next({ headers });
}

export const config = {
  matcher: [
    "/((?!api|_next/static|_next/image|favicon.ico).*)",
  ],
};
```

It would be easier if we could do it directly in the middleware, but this is a good workaround. Granted it only showcases protecting route groups, but it can be adapted to most use cases.

**Comment by jesper-bylund** [comment]
+1

**Comment by caipicoder** [comment]
+1

**Comment by JordanHoffman** [comment]
+1

**Comment by oswaldoapros** [comment]
+1

**Comment by luigiremor** [comment]
+1

**Comment by elliotBraem** [comment]
+1

**Comment by Dezmonter** [comment]
+1

**Comment by MaximPixel** [comment]
+1

**Comment by ViktorPls** [comment]
+1

**Comment by groyDoorvest** [comment]
+1
Knowing the Vercel team, if this is taking time, it must be hard to implement.
Still, this would be an insane DX improvement!

**Comment by VasilKostadinov0104** [comment]
+1

**Comment by betgo** [comment]
+1

**Comment by bmax** [comment]
:+1:

**Comment by KBTjunior** [comment]
+1

**Comment by anindosarker** [comment]
Hey everyone, since its taking so long to implement, why don't we start implementing it on our own?

**Reply by elliotBraem** [reply]
I’ve opted for tanstack router. So much cleaner

I recommend anyone else to do the same

**Reply by anindosarker** [reply]
you use tanstack router in nextjs?

**Reply by elliotBraem** [reply]
Oh no, I’ve stopped using next.js

**Comment by ftognetto** [comment]
+1

**Comment by EmmettM** [comment]
+1

**Comment by Dezmonter** [comment]
+100500

**Comment by richardpicucci** [comment]
+1
I understand you can do a match but when you're working in large organizations with many developers, you're requiring each developer to make sure they add their url to the authenticate/unauthenticated group (example).  I have 2 root folders like other people.  (authentication)  and (public).  I need to redirect like other people but at the moment, I have to match a url from the public group (urls).  As things evolve over time and new developers come onboard, currently, they are required to know/remember to add their public urls to the "match" middleware....  Yes, I can do things like run a script to programmatically build the url list  but I like the group route concept but we nee to be able to identify on the server, when the app router will resolve to a group route without checking each route.

**Comment by luizrobertomeier** [comment]
+1

**Comment by anton000** [comment]
This probably wont get implemented soon, as it would require a significant architectural change. A quick hack cant even be dome because route groups don’t currently exist in any runtime outputs or manifests. 😔

**Comment by hosseinipour** [comment]
+1

**Comment by GSGustavo** [comment]
+1

**Comment by AbianS** [comment]
+1

**Comment by Wonkers0** [comment]
+1

**Comment by RKHarrison** [comment]
+1

**Comment by Geancarlo-Cojocaru** [comment]
For anyone still looking for a solution to this: I’ve shared a small utility I wrote that generates route maps for Next.js projects.
I originally built this for our internal workflow at work. I preferred the copy/paste variant because each project has its own specificities, and I wanted it to be easy to refactor or customize. Hopefully, this saves some of you some manual work!

Repo: https://github.com/Geancarlo-Cojocaru/next-route-maps
